papis/isbn.py

```diff
@@ -11,24 +17,256 @@
+    meta["Authors"] = volumeinfo.get("authors", [""])
+    meta["Publisher"] = volumeinfo.get("publisher", "").strip('"')
+    if "publishedDate" in volumeinfo and len(volumeinfo["publishedDate"]) >= 4:
+        meta["Year"] = volumeinfo["publishedDate"][0:4]
+    else:
+        meta["Year"] = ""
+    meta["Language"] = volumeinfo.get("language", "")
+    return meta
+
+
+def meta_openl(isbn: str) -> dict[str, Any] | None:
+    url = (
+        f"https://openlibrary.org/api/books.json?bibkeys=ISBN:{isbn}"
+        "&jscmd=data"
+    )
+    data = json_request(url)
+    if not data:
+        return None
+    try:
+        volumeinfo = data[f"ISBN:{isbn}"]
+    except KeyError:
+        return None
+    meta: dict[str, Any] = {}
+    meta["ISBN-13"] = isbn
+    title = volumeinfo.get("title", "").replace(" :", ":")
+    subtitle = volumeinfo.get("subtitle", "")
+    meta["Title"] = format_title(title, subtitle)
+    meta["Authors"] = [
+        a["name"] for a in volumeinfo.get("authors", [{"name": ""}])
+    ]
+    meta["Publisher"] = volumeinfo.get("publishers", [{"name": ""}])[0]["name"]
+    if date := volumeinfo.get("publish_date", ""):
+        match = re.search(r"\d{4}", date)
+        meta["Year"] = match.group(0) if match else ""
+    else:
+        meta["Year"] = ""
+    # It seems like open library does not provide the language
+    meta["Language"] = ""
+    return meta
+
+
+def meta_wiki(isbn: str) -> dict[str, Any] | None:
+    url = (
+        f"https://en.wikipedia.org/api/rest_v1/data/citation/mediawiki/{isbn}"
+    )
+    data = json_request(url)
+    if not data:
+        return None
+    try:
+        volumeinfo = data[0]
+    except KeyError:
+        return None
+    meta: dict[str, Any] = {}
+    meta["ISBN-13"] = isbn
+    title = volumeinfo.get("title", "").replace(" :", ":")
+    meta["Title"] = format_title(title)
+    meta["Authors"] = [" ".join(a) for a in volumeinfo.get("author", [""])]
+    if not meta["Authors"] or len(meta["Authors"]) == 0:
+        meta["Authors"] = [
+            " ".join(a) for a in volumeinfo.get("contributor", [""])
+        ]
+    meta["Publisher"] = volumeinfo.get("publisher", "")
+    if date := volumeinfo.get("date", ""):
+        match = re.search(r"\d{4}", date)
+        meta["Year"] = match.group(0) if match else ""
+    else:
+        meta["Year"] = ""
+    # It seems like wiki does not provide the language
+    meta["Language"] = ""
+    return meta
 
 
-def get_data(query: str = "",
-             service: str | None = None) -> list[dict[str, Any]]:
+def meta_fallback(isbn: str | None, service: str) -> dict[str, Any] | None:
+    if not isbn or notisbn_fallback(isbn):
+        return None
+    if service == "goob":
+        return meta_goob(isbn)
+    elif service == "openl":
+        return meta_openl(isbn)
+    elif service == "wiki":
+        return meta_wiki(isbn)
+    else:
+        return None
+
+
+try:
+    from isbnlib import notisbn
+
+    __all__ = ["notisbn"]
```

Comment:
Other functions are also imported from this file. Is this necessary for something?

Reply:
Hm, is this because of `mypy` or something complaining about an implicit export? It should be fine to just add a `type: ignore` or `noqa` in that case.